content/src/logic/database-queries/snapshots-queries.ts

```diff
@@ -33,3 +35,182 @@ export async function* streamActiveDeployments(
+    return {
+      hash,
+      timeRange: {
+        initTimestamp,
+        endTimestamp
+      },
+      replacedSnapshotHashes,
+      numberOfEntities
+    }
+  })
+}
+
+export async function saveSnapshot(
+  database: AppComponents['database'],
+  snapshotMetadata: NewSnapshotMetadata,
+  generationTimestampSecs: number
+): Promise<void> {
+  const query = SQL`
+  INSERT INTO snapshots
+  (hash, init_timestamp, end_timestamp, replaced_hashes, number_of_entities, generation_time)
+  VALUES
+  (
+    ${snapshotMetadata.hash},
+    to_timestamp(${snapshotMetadata.timeRange.initTimestamp} / 1000.0),
+    to_timestamp(${snapshotMetadata.timeRange.endTimestamp} / 1000.0),
+    ${snapshotMetadata.replacedSnapshotHashes ?? []},
+    ${snapshotMetadata.numberOfEntities},
+    to_timestamp(${generationTimestampSecs} / 1000.0)
+  )
+  RETURNING hash
+  `
+  await database.queryWithValues(query, 'save_snapshot')
+}
+
+export async function getSnapshotHashesNotInTimeRange(
+  database: AppComponents['database'],
+  snapshotHashes: string[],
+  timeRange: TimeRange
+): Promise<Set<string>> {
+  if (snapshotHashes.length == 0) {
+    return new Set()
+  }
+  const query = SQL`
+  SELECT hash
+  FROM snapshots
+  WHERE init_timestamp <= end_timestamp
+  AND (init_timestamp >= to_timestamp(${timeRange.endTimestamp} / 1000.0)
+  OR end_timestamp <= to_timestamp(${timeRange.initTimestamp} / 1000.0))
+  AND hash IN (`
+  const hashes = snapshotHashes.map((h, i) => (i < snapshotHashes.length - 1 ? SQL`${h},` : SQL`${h}`))
+  hashes.forEach((hash) => query.append(hash))
+  query.append(`);`)
+
+  const result = await database.queryWithValues<{ hash: string }>(query, 'get_snapshots')
+  return new Set(result.rows.map((row) => row.hash))
+}
+
+export async function deleteSnapshotsInTimeRange(
+  database: AppComponents['database'],
+  snapshotHashesToDelete: string[],
+  timeRange: TimeRange
+): Promise<void> {
+  if (snapshotHashesToDelete.length == 0) {
+    return
+  }
+  const query = SQL`
+    DELETE FROM snapshots
+    WHERE init_timestamp >= to_timestamp(${timeRange.initTimestamp} / 1000.0)
+    AND end_timestamp <= to_timestamp(${timeRange.endTimestamp} / 1000.0)
+    AND hash IN (`
+  const hashes = snapshotHashesToDelete.map((h, i) => (i < snapshotHashesToDelete.length - 1 ? SQL`${h},` : SQL`${h}`))
+  hashes.forEach((hash) => query.append(hash))
+  query.append(`);`)
+  await database.queryWithValues(query, 'save_snapshot')
+}
+
+export async function saveProcessedSnapshot(
+  database: AppComponents['database'],
+  processedSnapshotHash: string,
+  processTimestampSecs: number
+): Promise<void> {
+  const query = SQL`
+  INSERT INTO processed_snapshots
+  (hash, process_time)
+  VALUES
+  (${processedSnapshotHash}, to_timestamp(${processTimestampSecs} / 1000.0))
+  RETURNING hash
+  `
+  await database.queryWithValues(query, 'save_processed_snapshot')
+}
+
+export async function deleteProcessedSnapshots(
```

Comment:
why you should ever need to delete a processed snapshot?

Reply:
when you're replacing 7 daily snapshots for a weekly snapshot for example